azkaban-common/src/main/java/azkaban/project/JdbcProjectImpl.java

```diff
@@ -802,6 +804,89 @@ public void uploadProjectProperties(final Project project,
     }
   }
 
+  @Override
+  public synchronized FlowRecommendation createFlowRecommendation(final int projectId, final String flowId)
+      throws ProjectManagerException {
+    logger.info("Creating flow recommendation " + flowId);
+    final String INSERT_FLOW_RECOMMENDATION =
+        "INSERT INTO project_flow_recommendations (project_id, flow_id, modified_time) values (?,?,?)";
+
+    final SQLTransaction<Integer> insertFlowRecommendation = transOperator ->
+        transOperator.update(INSERT_FLOW_RECOMMENDATION, projectId, flowId, System.currentTimeMillis());;
+
+    // Insert flow recommendation
+    try {
+      final int numRowsInserted = this.dbOperator.transaction(insertFlowRecommendation);
+      if (numRowsInserted == 0) {
+        throw new ProjectManagerException("No flow recommendations have been inserted.");
+      }
+    } catch (final SQLException ex) {
+      logger.error(INSERT_FLOW_RECOMMENDATION + " failed.", ex);
+      throw new ProjectManagerException("Insert flow recommendation " + flowId + " for existing "
+          + "project failed. ", ex);
+    }
+    return fetchFlowRecommendation(projectId, flowId);
+  }
+
+  @Override
+  public void updateFlowRecommendation(final FlowRecommendation flowRecommendation)
+      throws ProjectManagerException {
+    logger.info("Updating flow recommendation " + flowRecommendation.getId());
+    final String UPDATE_FLOW =
+        "UPDATE project_flow_recommendations SET cpu_recommendation=?,memory_recommendation=?,"
+            + "disk_recommendation=?,modified_time=? WHERE id=?";
+    try {
+      this.dbOperator
+          .update(UPDATE_FLOW, flowRecommendation.getCpuRecommendation(),
+              flowRecommendation.getMemoryRecommendation(),
+              flowRecommendation.getDiskRecommendation(), System.currentTimeMillis(),
+              flowRecommendation.getId());
+    } catch (final SQLException e) {
+      logger.error("Error updating flow recommendation", e);
+      throw new ProjectManagerException("Error updating flow recommendation " + flowRecommendation.getId(), e);
+    }
+  }
+
+  @Override
+  public FlowRecommendation fetchFlowRecommendation(final int projectId, final String flowId)
+      throws ProjectManagerException {
+    logger.info("Fetching flow recommendation. ProjectId: " + projectId + ", FlowId: " + flowId);
+    try {
+      return this.dbOperator
+          .query(ProjectFlowRecommendationsResultHandler.SELECT_PROJECT_FLOW_RECOMMENDATION,
+              new ProjectFlowRecommendationsResultHandler(), projectId, flowId).get(0);
+    } catch (final SQLException e) {
+      logger.error("Error fetching flow recommendation", e);
+      throw new ProjectManagerException("Error fetching flow recommendation. ProjectId: " + projectId + ", FlowId: " + flowId, e);
+    }
+  }
+
+  @Override
+  public List<FlowRecommendation> fetchAllProjectFlowRecommendations(final Project project) throws ProjectManagerException {
```

Comment:
I think actually fetchAllProjectFlowRecommendations and fetchAllFlowRecommendationsForProjects could have consist behavior. This fetchAllProjectFlowRecommendations could have return null, if there isn't a match for given project; while fetchAllFlowRecommendationsForProjects always return notNull even if not a match result. Either both null or both empty so it easier to use them.
Also adding comments for each method with returning values with empty/null/list of results in the interface, and with Exceptions.

Reply:
Hmmm, I found fetchAllProjectFlowRecommendations is not used at all: removed.